sql/catalyst/src/main/scala/org/apache/spark/sql/catalyst/expressions/collectionOperations.scala

```diff
@@ -287,3 +290,191 @@ case class ArrayContains(left: Expression, right: Expression)
+    val variableName = ctx.freshName("numElements")
+    val code = (0 until children.length)
+      .map(idx => s"$variableName += args[$idx].numElements();")
+      .foldLeft(s"int $variableName = 0;")((acc, s) => acc + "\n" + s)
+    (code, variableName)
+  }
+
+  private def nullArgumentProtection() : String = {
+    children.zipWithIndex
+      .filter(_._1.nullable)
+      .map(ci => s"if (args[${ci._2}] == null) return null;")
+      .mkString("\n")
+  }
+
+  private def genCodeForPrimitiveArrayConcat(ctx: CodegenContext, elementType: DataType): String = {
+    val arrayName = ctx.freshName("array")
+    val arraySizeName = ctx.freshName("size")
+    val counter = ctx.freshName("counter")
+    val arrayDataName = ctx.freshName("arrayData")
+
+    val (numElemCode, numElemName) = genCodeForNumberOfElements(ctx)
+
+    val unsafeArraySizeInBytes = s"""
+      |int $arraySizeName = UnsafeArrayData.calculateHeaderPortionInBytes($numElemName) +
+      |${classOf[ByteArrayMethods].getName}.roundNumberOfBytesToNearestWord(
+      |  ${elementType.defaultSize} * $numElemName
+      |);
+      """.stripMargin
+    val baseOffset = Platform.BYTE_ARRAY_OFFSET
+
+    val primitiveValueTypeName = CodeGenerator.primitiveTypeName(elementType)
+    val assignments = (0 until children.length).map { idx =>
+      s"""
+         |for (int z = 0; z < args[$idx].numElements(); z++) {
+         |  if (args[$idx].isNullAt(z)) {
+         |    $arrayDataName.setNullAt($counter);
+         |  } else {
+         |    $arrayDataName.set$primitiveValueTypeName(
+         |      $counter,
+         |      args[$idx].get$primitiveValueTypeName(z)
+         |    );
+         |  }
+         |  $counter++;
+         |}
+        """.stripMargin
+    }.mkString("\n")
+
+    s"""new Object() {
+       |  public ArrayData concat(${CodeGenerator.javaType(dataType)}[] args) {
+       |    ${nullArgumentProtection()}
+       |    $numElemCode
+       |    $unsafeArraySizeInBytes
+       |    byte[] $arrayName = new byte[$arraySizeName];
+       |    UnsafeArrayData $arrayDataName = new UnsafeArrayData();
+       |    Platform.putLong($arrayName, $baseOffset, $numElemName);
+       |    $arrayDataName.pointTo($arrayName, $baseOffset, $arraySizeName);
+       |    int $counter = 0;
+       |    $assignments
+       |    return $arrayDataName;
+       |  }
+       |}""".stripMargin
+  }
+
+  private def genCodeForComplexArrayConcat(ctx: CodegenContext): String = {
+    val genericArrayClass = classOf[GenericArrayData].getName
+    val arrayName = ctx.freshName("arrayObject")
+    val counter = ctx.freshName("counter")
+
+    val (numElemCode, numElemName) = genCodeForNumberOfElements(ctx)
+
+    val assignments = (0 until children.length).map { idx =>
+      s"""
+         |for (int z = 0; z < args[$idx].numElements(); z++) {
+         |  $arrayName[$counter] = args[$idx].array()[z];
+         |  $counter++;
+         |}
+        """.stripMargin
+    }.mkString("\n")
```

Comment:
To use `mkString` may lead to a compilation error due to 64KB bytecode limitation. Would it be possible to use `CodegenContext.splitExpressions()`?